Source/Scene/Cesium3DTilesetTraversal.js

```diff
@@ -0,0 +1,823 @@
+        // There may also be a tight box around just the tile's contents, e.g., for a city, we may be
+        // zoomed into a neighborhood and can cull the skyscrapers in the root node.
+        if (tile.contentReady && (
+                (tile.visibilityPlaneMask === CullingVolume.MASK_INSIDE) ||
+                (tile.contentVisibility(frameState) !== Intersect.OUTSIDE)
+            )) {
+            tileset._selectedTiles.push(tile);
+
+            var tileContent = tile.content;
+            if (tileContent.featurePropertiesDirty) {
+                // A feature's property in this tile changed, the tile needs to be re-styled.
+                tileContent.featurePropertiesDirty = false;
+                tile.lastStyleTime = 0; // Force applying the style to this tile
+                tileset._selectedTilesToStyle.push(tile);
+            }  else if ((tile.lastSelectedFrameNumber !== frameState.frameNumber - 1)) {
+                // Tile is newly selected; it is selected this frame, but was not selected last frame.
+                tileset._selectedTilesToStyle.push(tile);
+            }
+            tile.lastSelectedFrameNumber = frameState.frameNumber;
+        }
+    }
+
+    // PERFORMANCE_IDEA: is it worth exploiting frame-to-frame coherence in the sort, i.e., the
+    // list of children are probably fully or mostly sorted unless the camera moved significantly?
+    function sortChildrenByDistanceToCamera(a, b) {
+        // Sort by farthest child first since this is going on a stack
+        if (b.distanceToCamera === 0 && a.distanceToCamera === 0) {
+            return b._centerZDepth - a._centerZDepth;
+        }
+
+        return b.distanceToCamera - a.distanceToCamera;
+    }
+
+    var emptyArray = freezeObject([]);
+
+    function BaseTraversal() {
+        this.tileset = undefined;
+        this.frameState = undefined;
+        this.outOfCore = undefined;
+        this.stack = new ManagedArray();
+        this.leaves = new ManagedArray();
+        this.baseScreenSpaceError = undefined;
+        this.internalDFS = new InternalBaseTraversal();
+    }
+
+    BaseTraversal.prototype.execute = function(tileset, root, baseScreenSpaceError, frameState, outOfCore) {
+        this.tileset = tileset;
+        this.frameState = frameState;
+        this.outOfCore = outOfCore;
+        this.leaves.length = 0;
+        this.baseScreenSpaceError = Math.max(baseScreenSpaceError, this.tileset._maximumScreenSpaceError);
+        this.internalDFS.tileset = this.tileset;
+        this.internalDFS.frameState = this.frameState;
+        this.internalDFS.outOfCore = this.outOfCore;
+        this.internalDFS.baseScreenSpaceError = this.baseScreenSpaceError;
+        DFS(root, this);
+    };
+
+    BaseTraversal.prototype.visitStart = function(tile) {
+        if (tile._lastVisitedFrame !== this.frameState.frameNumber) {
+            visitTile(this.tileset, tile, this.frameState, this.outOfCore);
+        }
```

Comment:
This last visited check is pretty much the same throughout, why not put back in `visitTile`?